doc/design.md

```diff
@@ -0,0 +1,352 @@
+   libtorch with github.com/pytorch/xla, we get an additional implementation for
+   Google TPU.
+
+1. A higher-level abstraction is in the Python package torch.nn.functional,
+   which provides functions defined in Python and calls native functions in
+   C/C++.
+
+1. The highest layer is modules; each is a Python class with a method forward
+   defining the forward computation process and data members that can store
+   states.
+
+## Tensors and Garbage Collection
+
+libtorch includes the C++ definition of the fundamental data type `at::Tensor`
+and native functions that operate it.
+
+The key design feature of the tensor is automatic garbage collection (GC). In
+C++, the class `at::Tensor` contains only one data member, a smart pointer to
+the real tensor object.  This smart pointer performs reference count-based GC,
+which frees a tensor once its reference count gets zero.  Comparing with Go and
+Java’s GC, which runs the mark-and-sweep algorithm, reference count reacts
+instantly but cannot handle the case of cyclic-dependency.
+
+PyTorch programmers access `at::Tensor` from the Python binding. Python’s GC
+uses reference counts like `shared_ptr`.  For completeness, Python runs
+mark-and-sweep from time to time to handle cyclic-dependencies.
+
+Go provides an asynchronous API, `runtime.GC()`, to trigger GC and returns
+immediately without waiting for the completion of GC.  If all tensors are in CPU
+memory, this mechanism works; however, in deep learning, we would prefer to host
+tensors in GPU memory, which is a precious resource. We prefer to free tensors
+immediately when they are out-of-use so that the next iteration can create new
+tensors in GPU.
+
+### Synchronize Go’s GC
+
+We started with inventing new GC mechanisms in the library, including adding a
+global reference count table. However, after trying several strategies, we
+noticed that we could customize Go’s GC for the tensor type in GoTorch
+specifically to make it synchronous, or able to wait till the completion of GC
+before returning.
+
+The basic idea behind the design is the categorization of tensors by different
+purposes in deep learning:
+
+1. model parameters -- created before, updated during, and freed after the train
+   loop,
+1. buffers -- with lifespan similar to model parameters but used to BatchNorm to
+   keep statistics of input data, and
+1. intermediate results -- including those generated during the forward and
+   backward pass in each step of the train loop.
+
+The customized Go GC mechanism doesn’t handle the first two categories, which is
+the topic of the next section, Porting Modules.
+
+To handle intermediate results, GoTorch users need to call `gotorch.GC()` at the
+beginning of each train loop step.  The first job of `gotorch.GC()` is to mark
+that all tensors generated since then, which are considered intermediate
+results, are subject to the customized GC.  After the train loop, users are
+supposed to call `gotorch.FinishGC()` to unset the mark.
```

Comment:
why do users need to call `FinishGC`? All masked tensors should have been recycled when exiting loops?

Reply:
The function `torch.GC` is supposed to be invoked at the beginning of an iteration, as it enables the marking of subsequently created tensors to be intermediate results.

It might be more comprehensive to write

```go
torch.BeginGC()
for ... {
    // the train loop
    torch.GC()
}
torch.EndGC()
```